config: Add SerializableRegex to serialize Regex (#1307)

marianomarciello opened the pull request:
# Description

 SerializableRegex allows to serialize Regex without using a reference to an option. This will avoid the clippy warning about ref_option.

Fixes #(1307)

## Type of change

- [ ] Development change (no change visible to user)
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation only update (no change to the factual codebase)
- [ ] This change requires a documentation update

## Updated user documentation:

Please insert user documentation that should be updated (as in the wiki).

See [CONTRIBUTING.md User Documentation section](../CONTRIBUTING.md#user-documentation) for further details.

**Note: Manual page changes must be performed in a commit, not in this PR section.**

# Checklist:

- [x ] Ran `make test` locally with no errors or warnings reported
  Note: To fully reproduce CI checks, you will need to run `make test-full-nix`. Usually, this is not necessary.
- [ ] Manual page has been updated accordingly
- [ ] Wiki pages have been updated accordingly (to perform **after** merge)

mautamu commented on leftwm/src/config.rs:
```diff
@@ -96,6 +86,46 @@ pub struct WindowHook {
+impl SerializableRegex {
+    /// # Panics
+    ///
+    /// Function can only panic if the str used is not a valid regex
```
Probably should call out for reference which syntax to use; think it's still https://github.com/google/re2/wiki/Syntax, no?

marianomarciello replied:
I think that maybe instead of using a github page with the definition, it might be easier to link the docs from the create that is used ? https://docs.rs/regex/latest/regex/#syntax

mautamu replied:
Works even better! Great idea.

mautamu commented on leftwm/src/config.rs:
```diff
@@ -96,6 +86,46 @@ pub struct WindowHook {
+    #[must_use]
+    pub fn new(str: &str) -> Self {
+        SerializableRegex {
+            regex: Regex::new(str)
```
Since we can return an Error from deserialize(), should we also do so here and use a chained let in deserialize() to surface the error upwards without a panic?

marianomarciello replied:
Added in the below commit. :)

mautamu replied:
Awesome, thank you very much! :)

mautamu reviewed (commented):
These changes seem reasonable to me, though yet to test. Left a couple comments.

marianomarciello commented:
Hey @VuiMuich  I have added two tests in my latest commit. Please if you have time take a look :). Thanks

VuiMuich commented:
Thanks @marianomarciello, looking good. 
Just a tiny nitpick and one thing I am unsure about myself:
- Should the first be actually better be two separate tests for serialization and deserialization, so each test is as minimal as possible? To be fair I am no expert on this and I am completely fine to leave this as is, too.
- it occurred to me, right after sending my previous remark, but hadn't time to edit and then forgot about it until now reading the ready test: the second test should be named `attempt_deserialize_...`

marianomarciello commented:
Hey @VuiMuich  I added two more tests and fixed the typo.
Thanks for spotting that :D 

Let me know if any other change for this PL are needed. Thanks.

VuiMuich commented:
Good thinking with the two new tests. And thanks for your contribution!
